sources/osgShader/ShaderGenerator.js

```diff
@@ -203,13 +215,17 @@ define( [
 
             return function ( state ) {
                 // extract valid attributes
-                var hash = '';
+
+                this._hashStack.current = 0;
```

Comment:
Good point about that array member added. I think I checked how we did it in cullStack but maybe I misread.

Array could be global too, why not indeed.

Rationale and results are explained in the pivotal task, but anyway:

tl;dr: getHash does pop regularly on js profiler, this patch does reduce by half GC firefox in the boxe: https://sketchfab.com/models/663db593a73e4ceabf21a457f5ebcb26 and was my guess on the matcap/shadeless thing (as it's the only real difference between the two)

it's called each frame, very often (the cache flatten lots of not "flattened out" stateset like it theoritically could), and it'll it's all stateattributes which in turn does string concat of it parameter each frame + all texture attributes stack (notice the "stack")